### PR TITLE
Fix transfermarket functions

### PR DESCRIPTION
**Pull request by tonyelhabr**
Swap `xml2::read_html()` with `.load_page()` in transfermarket functions.

It seems that transfermarket started blocking the default httr user agent for HTTP requests. The solution is to use the `.load_page()` function, which we developed when `{worldfootballR}` started getting blocked on fbref. `.load_page()` provides a non-default user agent that is based on the versions of RStudio, R, and the OS that the user has. For example, mine looks like this

```r
library(worldfootballR)
getOption("worldfootballR.agent")
#> [1] "RStudio Desktop (2023.3.1.446); R (4.2.0 x86_64-w64-mingw32 x86_64 x86_64)"
```

### CODE REVIEW

**Review by JaseZiv — COMMENTED** [review]
@tonyelhabr I had tried some of these options locally the other day and they also failed...

What I did find that worked though was the below...

Loading pages as we did with `.load_page()` failed with a 403 as under the hood, that function is still using xml2::read_html() after setting the user agent...
```
player_url <- "https://www.transfermarkt.com/eden-hazard/profil/spieler/50202"
player_page <- tryCatch(.load_page(player_url), error = function(e) NA)
```

While trying to read in the content like the below worked:

```
headers = c(
 `user-agent` = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 
 (KHTML, like Gecko) Chrome/102.0.5005.61 Safari/537.36'
)
    
res <- httr::GET(url = player_url, httr::add_headers(.headers=headers)) |> httr::content()
    
player_name <- res |> rvest::html_nodes("div h1")|> rvest::html_text() |> stringr::str_squish()
```

A few questions then;

1. Should we switch all `xml2::read_html()` calls to `httr::GET()` and pass in `httr::add_headers()` ?
2. how do we dynamically set `user-agent` based on the user's environment?

**Comment by tonyelhabr** [comment]
> @tonyelhabr I had tried some of these options locally the other day and they also failed...
> 
> What I did find that worked though was the below...
> 
> Loading pages as we did with `.load_page()` failed with a 403 as under the hood, that function is still using xml2::read_html() after setting the user agent...
> 
> ```
> player_url <- "https://www.transfermarkt.com/eden-hazard/profil/spieler/50202"
> player_page <- tryCatch(.load_page(player_url), error = function(e) NA)
> ```
> 
> While trying to read in the content like the below worked:
> 
> ```
> headers = c(
>  `user-agent` = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 
>  (KHTML, like Gecko) Chrome/102.0.5005.61 Safari/537.36'
> )
>     
> res <- httr::GET(url = player_url, httr::add_headers(.headers=headers)) |> httr::content()
>     
> player_name <- res |> rvest::html_nodes("div h1")|> rvest::html_text() |> stringr::str_squish()
> ```
> 
> A few questions then;
> 
>     1. Should we switch all `xml2::read_html()` calls to `httr::GET()` and pass in `httr::add_headers()` ?
> 
>     2. how do we dynamically set `user-agent` based on the user's environment?

Sorry, I thought just swapping in `.load_page()` would work, but it appears it doesn't.

1. Yes, I think that's probably a good solution. `httr::content()` will use `xml2::read_html()` for parsing, but we get the benefit of explicitly setting the user agent with `httr::add_headers()` in `httr::GET()`.
2. We currently set the `user-agent` dynamically at package load time (see `worldfootballR.agent` in `zzz.R`) based on the user's version of R, RStudio, and their OS. However,that string gets blocked, and, instead, we need a string that looks like a browser, like the one in your example and in the SO post.
    - I'm not sure there's a good, dynamic solution. Maybe we can get a string that looks like that by identifying what browser the user has on their machine? I'm not aware of how we could do that tho.
    - Maybe we can predefine a set of like 5 strings like the one in your example, 1 specifying Firefox + MacOS, another specifying Edge + Windows, etc. And then we randomly sample from the 5 strings at package load time, and set the `worldfootballR.agent` option in a similar fashion to the way it is done today.

File: R/internals.R [hunk 1]
```diff
@@ -400,18 +400,54 @@
...
+.load_page_tm <- function(page_url) {
+  ua_options <- c(
+    'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.103 Safari/537.36',
+    'Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:47.0) Gecko/20100101 Firefox/47.0',
+    'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36 Edg/91.0.864.59',
+    'Mozilla/5.0 (Macintosh; Intel Mac OS X x.y; rv:42.0) Gecko/20100101 Firefox/42.0',
+    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.5005.61 Safari/537.36'
+  )
+
+  ua <- sample(ua_options, 1)
```
Comment (JaseZiv) [inline R/internals.R:439]: Not the most intelligent solution to dynamically setting the user agent, but tested on macos locally and all values returned page data...

Thoughts on this @tonyelhabr?

**Comment by JaseZiv** [comment]
Looks like that solution didn't work for tests running on ubuntu using GHA...

Do we want to solve for the automated tests running using GHA, or push these changes through for users manually triggering requests? As I type though, I realise that any automated scripts run using GHA will fail too...

File: R/internals.R (hunk 1 above)
Comment (tonyelhabr) [reply R/internals.R]: It's not bad as a first pass. I found [this package](https://github.com/fangzhou-xie/Randomuseragent) that can randomly generate a valid user agent for us.

In fact, I've just pushed some changes that incorporate this package. The changes also consolidate `.load_page()` and `.load_page_tm()` into just `.load_page()`, always using the randomized browser user agent. The randomized browser user agent should be fine to use for non-transfermarket functions that, up to this point, were using `.load_page()` + the user-specific R desktop agent.

Comment (tonyelhabr) [reply R/internals.R]: It's not bad as a first pass. I found [this package](https://github.com/fangzhou-xie/Randomuseragent) that can randomly generate a valid user agent for us.

In fact, I've just pushed some changes that incorporate this package. The changes also consolidate `.load_page()` and `.load_page_tm()` into just `.load_page()`, always using the randomized browser user agent. The randomized browser user agent should be fine to use for non-transfermarket functions that, up to this point, were using `.load_page()` + the user-specific R desktop agent.

**Comment by tonyelhabr** [comment]
> Looks like that solution didn't work for tests running on ubuntu using GHA...
> 
> Various errors now including:
> 
> ```
> Error in `.[[1]]`: subscript out of bounds
> ```
> 
> and
> 
> ```
> <rlang_error_data_pronoun_not_found/rlang_error/error/condition>
> Error in `.data$X.`: Column `X.` not found in `.data`.
> ```
> 
> etc...
> 
> So I'm assuming `httr::GET()` is returning something, just not what we want?
> 
> Do we want to solve for the automated tests running using GHA, or push these changes through for users manually triggering requests? As I type though, I realise that any automated scripts run using GHA will fail too...

i'm seeing the same issue after pushing my changes. Tests pass on local, but not on GHA. I wonder if GHA is using package versions that are incompatible (e.g. a super new version of `httr` that behaves differently).

Maybe we can try being very specific with package versions in the DESCRIPTION file to see if we can get the tests to pass once, and then revert the version specifications once we've got a better understanding of things.

**Comment by JaseZiv** [comment]
Closing this for now as scrapers no longer being blocked.

Will create a new PR to get around 403s being returned when running tests through `devtools::check()`

**Pull request closed without merging**